Stargate fails to start because of too many open files

I am running Stargate 1.0.31 in Kubernetes with K8ssandra. The Stargate image used is stargateio/stargate-3_11:v1.0.31. In one of our automated tests we have seen Stargate fail to start a few times with this a resource limit error like this:

```console
INFO  [main] 2021-09-28 21:25:10,597 AbstractConnector.java:331 - Started Server@29862f53{HTTP/1.1, (http/1.1)}{0.0.0.0:8082}
INFO  [main] 2021-09-28 21:25:10,598 Server.java:415 - Started @80358ms
INFO  [main] 2021-09-28 21:25:10,599 BaseActivator.java:185 - Started restapi
Finished starting bundles.
Unexpected error: java.io.IOException: User limit of inotify instances reached or too many open files
java.lang.RuntimeException: java.io.IOException: User limit of inotify instances reached or too many open files
        at io.stargate.starter.Starter.watchJarDirectory(Starter.java:539)
        at io.stargate.starter.Starter.start(Starter.java:441)
        at io.stargate.starter.Starter.cli(Starter.java:619)
        at io.stargate.starter.Starter.main(Starter.java:660)
Caused by: java.io.IOException: User limit of inotify instances reached or too many open files
        at sun.nio.fs.LinuxWatchService.<init>(LinuxWatchService.java:64)
        at sun.nio.fs.LinuxFileSystem.newWatchService(LinuxFileSystem.java:47)
        at io.stargate.starter.Starter.watchJarDirectory(Starter.java:526)
        ... 3 more
```

This is in a CI environment with limited cpu/memory resources. The test is running in the free tier runner in GitHubActions. The runner vm has 2 cpus and 7 GB memory. The particular test in which this failed had already deployed two Cassandra nodes and one Stargate node. This failure is from the second Stargate node.

I believe the open file limit on the vm is set to 65536. I don't think I am able to increase it. Maybe the solution is to run my tests in an environment with more resources, but it would be nice if Stargate could less demanding especially considering this happens on startup.

--- a/starctl
+++ b/starctl
@@ -155,6 +155,10 @@ if [[ ${#STARGATE_ARGS[@]} -eq 0 ]]; then
     if [[ ! -z "$DISABLE_MBEAN_REGISTRATION" ]]; then
         STARGATE_ARGS+=("--disable-mbean-registration")
     fi
+
+    if [[ ! -z "$DISABLE_BUNDLES_WATCH" ]]; then
+        STARGATE_ARGS+=("--disable-bundles-watch"")
+    fi
 fi
 
 declare -a CMD=(java -server)

--- a/starctl
+++ b/starctl
@@ -157,7 +157,7 @@ if [[ ${#STARGATE_ARGS[@]} -eq 0 ]]; then
     fi
 
     if [[ ! -z "$DISABLE_BUNDLES_WATCH" ]]; then
-        STARGATE_ARGS+=("--disable-bundles-watch"")
+        STARGATE_ARGS+=("--disable-bundles-watch")
     fi
 fi
 

--- a/stargate-starter/src/main/java/io/stargate/starter/Starter.java
+++ b/stargate-starter/src/main/java/io/stargate/starter/Starter.java
@@ -233,6 +233,14 @@ public class Starter {
       name = {"--disable-mbean-registration", "Whether the mbean registration should be disabled"})
   protected boolean disableMBeanRegistration = false;
 
+  @Order(value = 21)
+  @Option(
+      name = {
+        "--disable-bundles-watch",
+        "Whether watching the bundle directory for new jars load should be disabled"
+      })
+  protected boolean disableBundlesWatch = false;
+
   @Order(value = 1000)
   @Option(
       name = "--nodetool",
@@ -438,7 +446,9 @@ public class Starter {
 
     System.out.println(STARTED_MESSAGE);
 
-    if (watchBundles) watchJarDirectory(JAR_DIRECTORY);
+    if (watchBundles && !disableBundlesWatch) {
+      watchJarDirectory(JAR_DIRECTORY);
+    }
   }
 
   protected Map<String, String> felixConfig() {
@@ -536,7 +546,9 @@ public class Starter {
         poll = pollEvents(watchService);
       }
     } catch (IOException | ClosedWatchServiceException | InterruptedException e) {
-      throw new RuntimeException(e);
+      // Since most deployments will occur in the cloud its not worth breaking the start up to
+      // support hot-reload
+      System.err.printf("Jars will not be watched due to unexpected error: %s%n", e.getMessage());
     }
   }
 
@@ -624,10 +636,10 @@ public class Starter {
 
         // Parsing failed
         // Display errors and then the help information
-        System.err.println(String.format("%d errors encountered:", result.getErrors().size()));
+        System.err.printf("%d errors encountered:%n", result.getErrors().size());
         int i = 1;
         for (ParseException e : result.getErrors()) {
-          System.err.println(String.format("Error %d: %s", i, e.getMessage()));
+          System.err.printf("Error %d: %s%n", i, e.getMessage());
           i++;
         }
 
@@ -644,12 +656,12 @@ public class Starter {
         // a user just uses -h/--help
       } else {
         System.err.println();
-        System.err.println(String.format("Usage error: %s", p.getMessage()));
+        System.err.printf("Usage error: %s%n", p.getMessage());
         System.err.println();
       }
     } catch (Exception e) {
       // Errors should be being collected so if anything is thrown it is unexpected
-      System.err.println(String.format("Unexpected error: %s", e.getMessage()));
+      System.err.printf("Unexpected error: %s%n", e.getMessage());
       e.printStackTrace(System.err);
 
       System.exit(1);

--- a/stargate-starter/src/main/java/io/stargate/starter/Starter.java
+++ b/stargate-starter/src/main/java/io/stargate/starter/Starter.java
@@ -237,7 +237,7 @@ public class Starter {
   @Option(
       name = {
         "--disable-bundles-watch",
-        "Whether watching the bundle directory for new jars load should be disabled"
+        "Whether watching the bundle directory for new jars to load should be disabled"
       })
   protected boolean disableBundlesWatch = false;
 

